### PR TITLE
[66_5] Perf tuning on tm_reader::decode via index_of

### PR DESCRIPTION
**Pull request by darcy-shen**
## Why you open this Pull Request?
as title describe

## What work have you done in the current Pull Request?
For 5m tm doc: `564ms -> 543ms`

### CODE REVIEW

File: src/Data/Convert/Texmacs/fromtm.cpp [hunk 1]
```diff
@@ -76,9 +76,12 @@ tm_reader<format_without_utf8>::skip_blank () {
...
-  int    i, n= N (s);
-  string r;
-  for (i= 0; i < n; i++)
+  int start= index_of (s, '\\');
+  if (start == -1) return s;
+
+  int    n= N (s);
+  string r= s (0, start);
+  for (int i= start; i < n; i++)
     if (((i + 1) < n) && (s[i] == '\\')) {
```
Comment (jingkaimori) [inline src/Data/Convert/Texmacs/fromtm.cpp:85]:
```suggestion

  int    n= N (s);
  int start= index_of (s, '\\');
  if (start == -1 || start == n-1) return s;

  string r= s (0, start);
  for (int i= start; i < n; i++)
    if (s[i] == '\\') {
```

**Review by jingkaimori — CHANGES_REQUESTED** [review]
further optimization can be done, and qtest has facilities for benchmark:

https://doc.qt.io/qt-6/qttestlib-tutorial5-example.html

**Comment by darcy-shen** [comment]
> and qtest has facilities for benchmark:

we do no need micro bench here.

File: src/Data/Convert/Texmacs/fromtm.cpp (hunk 1 above)
Comment (darcy-shen) [reply src/Data/Convert/Texmacs/fromtm.cpp]: 这个代码有问题，原本的代码是正确的

Comment (darcy-shen) [reply src/Data/Convert/Texmacs/fromtm.cpp:85]: s里面如果有两个\，然后第二个\正好在最后，那么会出现i++之后i==n，这样数组越界了。

**Comment by darcy-shen** [comment]
Tested on Linux, there is no significant performance improvements.

File: src/Data/Convert/Texmacs/fromtm.cpp [hunk 2]
```diff
@@ -76,10 +76,13 @@ tm_reader<format_without_utf8>::skip_blank () {
...
-  int    i, n= N (s);
-  string r;
-  for (i= 0; i < n; i++)
-    if (((i + 1) < n) && (s[i] == '\\')) {
+  int start= index_of (s, '\\');
+  if (start == -1) return s;
+
+  int    n= N (s), n_minus_one= n - 1, i= start;
+  string r= s (0, start);
+  for (; i < n_minus_one; i++)
```
Comment (darcy-shen) [inline src/Data/Convert/Texmacs/fromtm.cpp:84]: 直接用n-1就可以了

File: src/Data/Convert/Texmacs/fromtm.cpp [hunk 3]
```diff
@@ -92,6 +95,10 @@ tm_reader<format_without_utf8>::decode (string s) {
       else r << s[i];
     }
     else r << s[i];
+  if (i < n) {
```
Comment (darcy-shen) [inline src/Data/Convert/Texmacs/fromtm.cpp:98]: 这里需要说明一下为什么要这样，增加一个单元测试覆盖 i=n 这种情况

**Comment by jingkaimori** [comment]
Performance is slightly improved:

Before:
```
PASS   : TestConverter::test_texmacs_to_tree_bench(29_1_1.tm)
RESULT : TestConverter::test_texmacs_to_tree_bench():"29_1_1.tm":
     0.018 msecs per iteration (total: 76, iterations: 4096)
PASS   : TestConverter::test_texmacs_to_tree_bench(46_3.tm)
RESULT : TestConverter::test_texmacs_to_tree_bench():"46_3.tm":
     0.394 msecs per iteration (total: 101, iterations: 256)
PASS   : TestConverter::test_texmacs_to_tree_bench(64_1.tm)
RESULT : TestConverter::test_texmacs_to_tree_bench():"64_1.tm":
     1.5 msecs per iteration (total: 98, iterations: 64)
```

After:
```
PASS   : TestConverter::test_texmacs_to_tree_bench(29_1_1.tm)
RESULT : TestConverter::test_texmacs_to_tree_bench():"29_1_1.tm":
     0.016 msecs per iteration (total: 69, iterations: 4096)
PASS   : TestConverter::test_texmacs_to_tree_bench(46_3.tm)
RESULT : TestConverter::test_texmacs_to_tree_bench():"46_3.tm":
     0.37 msecs per iteration (total: 95, iterations: 256)
PASS   : TestConverter::test_texmacs_to_tree_bench(64_1.tm)
RESULT : TestConverter::test_texmacs_to_tree_bench():"64_1.tm":
     1.4 msecs per iteration (total: 92, iterations: 64)
```

File: src/Data/Convert/Texmacs/fromtm.cpp (hunk 2 above)
Comment (jingkaimori) [reply src/Data/Convert/Texmacs/fromtm.cpp:84]: 这么写循环里还能少一条指令

**Review by jingkaimori — APPROVED** [review]
LGTM

**Comment by jingkaimori** [comment]
Modification should not be make on parser of tm. tmu parser needs these optimization.

**Pull request closed without merging**